File: android-app/src/main/java/com/medziku/plantwatering/bluetooth/BluetoothDeviceAdapter.java
package com.medziku.plantwatering.bluetooth;

import android.bluetooth.BluetoothDevice;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

import com.medziku.plantwatering.R;

import java.util.ArrayList;

public class BluetoothDeviceAdapter extends BaseAdapter {

    private Context mContext;
    private ArrayList<BluetoothDevice> mBoundedList;
    private ArrayList<BluetoothDevice> mAvailableList;

    private int mMyDeviceSeparatorPosition = 0;
    private int mBoundedDeviceSeparatorPosition = 0;
    private int mAvailableDeviceSeparatorPosition = 0;

    private String mMyDeviceSeparatorText, mBoundedDeviceSeparatorText,
            mAvailableDeviceSeparatorText;
    private String mDeviceName, mDeviceAddress;

    // View Type for Separators
    private static final int ITEM_VIEW_TYPE_SEPARATOR = 0;
    // View Type for Regular rows
    private static final int ITEM_VIEW_TYPE_REGULAR = 1;
    // Types of Views that need to be handled
    // -- Separators and Regular rows --
    private static final int ITEM_VIEW_TYPE_COUNT = 2;

    /**
     * Default log tag for this class.
     */
    private final String LOG_TAG = BluetoothDeviceAdapter.class.getSimpleName();

    public BluetoothDeviceAdapter(Context context, ArrayList<BluetoothDevice> boundedList,
                                  ArrayList<BluetoothDevice> availableList, String deviceName,
                                  String deviceAddress) {
        mContext = context;

        // Set array lists for bounded and available bluetooth devices
        mBoundedList = boundedList;
        mAvailableList = availableList;

        //Set section separator names
        mMyDeviceSeparatorText = context.getResources().
                getString(R.string.my_device);
        mBoundedDeviceSeparatorText = context.getResources().
                getString(R.string.bounded_devices);
        mAvailableDeviceSeparatorText = context.getResources().
                getString(R.string.available_devices);

        // Set this device bluetooth info
        mDeviceName = deviceName;
        mDeviceAddress = deviceAddress;
    }

    public BluetoothDeviceAdapter(Context context, String deviceName,
                                  String deviceAddress) {
        // Create with empty devices lists
        this(context, new ArrayList<BluetoothDevice>(),
                new ArrayList<BluetoothDevice>(), deviceName, deviceAddress);
    }

    @Override
    public int getCount() {

        // Calculate count. For each non empty device list add one for header.
        // Otherwise skip
        return 2 + (mBoundedList.size() == 0 ? 0 : (mBoundedList.size() + 1))
                + (mAvailableList.size() == 0 ? 0 : (mAvailableList.size() + 1));
    }

    @Override
    public Object getItem(int position) {
        return null;
    }

    @Override
    public long getItemId(int position) {
        return 0;
    }

    @Override
    public int getViewTypeCount() {
        return ITEM_VIEW_TYPE_COUNT;
    }

    @Override
    public int getItemViewType(int position) {
        if (position == mMyDeviceSeparatorPosition
                || position == mBoundedDeviceSeparatorPosition
                || position == mAvailableDeviceSeparatorPosition) {
            // If separator
            return ITEM_VIEW_TYPE_SEPARATOR;
        } else {
            // Regular row
            return ITEM_VIEW_TYPE_REGULAR;
        }
    }

    @Override
    public boolean isEnabled(int position) {
        // Disabled if separator
        return position != mMyDeviceSeparatorPosition
                && position != mBoundedDeviceSeparatorPosition
                && position != mAvailableDeviceSeparatorPosition;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        View view;

        int itemViewType = getItemViewType(position);

        if (convertView == null) {
            LayoutInflater inflater =(LayoutInflater) mContext.
                    getSystemService(Context.LAYOUT_INFLATER_SERVICE);

            if (itemViewType == ITEM_VIEW_TYPE_SEPARATOR) {
                // If separator
                view = inflater.inflate(R.layout.device_section_header, null);
            } else {
                // Regular row
                view = inflater.inflate(R.layout.device_item, null);
            }
        } else {
            view = convertView;
        }

        if (itemViewType == ITEM_VIEW_TYPE_SEPARATOR) {
            // If separator

            TextView separatorView = (TextView) view.findViewById(R.id.separator);
            if(position == mMyDeviceSeparatorPosition){
                separatorView.setText(mMyDeviceSeparatorText);
            }
            else if(position == mBoundedDeviceSeparatorPosition
                    && mBoundedDeviceSeparatorPosition != 0){
                separatorView.setText(mBoundedDeviceSeparatorText);
            }
            else if(position == mAvailableDeviceSeparatorPosition
                    && mAvailableDeviceSeparatorPosition != 0){
                separatorView.setText(mAvailableDeviceSeparatorText);
            }
        } else {
            // If regular row

            TextView deviceNameView = (TextView) view.findViewById(R.id.device_name);
            TextView deviceAddressView = (TextView) view.findViewById(R.id.device_address);

            // Set device name and address values
            if(position == 1){
                deviceNameView.setText(mDeviceName);
                deviceAddressView.setText(mDeviceAddress);
            }
            else {
                BluetoothDevice device;
                int positionInList = position - 3;
                if(positionInList < mBoundedList.size())
                    device = mBoundedList.get(positionInList);
                else{
                    positionInList -= mBoundedList.size() +1;
                    device = mAvailableList.get(positionInList);
                }

                deviceNameView.setText(device.getName());
                deviceAddressView.setText(device.getAddress());
            }
        }

        return view;
    }

    /**
     * Add new device to bounded devices list.
     * @param device new bluetooth device.
     */
    public void addBoundedDevice(BluetoothDevice device) {
        mBoundedList.add(device);
        updateSeparatorPositions();
        this.notifyDataSetChanged();
    }

    /**
     * Add new device to available devices list.
     * @param device new bluetooth device.
     */
    public void addAvailableDevice(BluetoothDevice device) {
        mAvailableList.add(device);
        this.notifyDataSetChanged();
    }

    /**
     * Recalculate separator positions
     */
    private void updateSeparatorPositions() {
        mBoundedDeviceSeparatorPosition = mBoundedList.size() == 0 ? 0 : 2;
        mAvailableDeviceSeparatorPosition = mAvailableList.size() == 0 ? 0 : mBoundedList.size()
                + mBoundedDeviceSeparatorPosition;
    }
}

File: android-app/src/main/kotlin/com/medziku/plantwatering/bluetooth/BluetoothDeviceAdapter.kt
package com.medziku.plantwatering.bluetooth

import android.bluetooth.BluetoothDevice
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.TextView
import com.medziku.plantwatering.R

class BluetoothDeviceAdapter(private val context: Context,
                             private val boundedList: ArrayList<BluetoothDevice> = ArrayList(),
                             private val availableList: ArrayList<BluetoothDevice> = ArrayList(),
                             private val deviceName: String,
                             private val deviceAddress: String) : BaseAdapter() {

    private val deviceSeparatorPosition = 0
    private var boundedDeviceSeparatorPosition = 0
    private var availableDeviceSeparatorPosition = 0

    private val deviceSeparatorText: String = context.resources.getString(R.string.my_device)
    private val boundedDeviceSeparatorText: String = context.resources.getString(R.string.bounded_devices)
    private val availableDeviceSeparatorText: String = context.resources.getString(R.string.available_devices)

    companion object {
        // View Type for Separators
        private const val ITEM_VIEW_TYPE_SEPARATOR = 0
        // View Type for Regular rows
        private const val ITEM_VIEW_TYPE_REGULAR = 1
        // Types of Views that need to be handled
        // -- Separators and Regular rows --
        private const val ITEM_VIEW_TYPE_COUNT = 2
    }

    override fun getCount(): Int {
        return 2 + (if (boundedList.size == 0) 0 else boundedList.size + 1) + (if (availableList.size == 0) 0 else availableList.size + 1)
    }

    override fun getItem(position: Int): Any? {
        return null
    }

    override fun getItemId(position: Int): Long {
        return 0
    }

    override fun getViewTypeCount(): Int {
        return ITEM_VIEW_TYPE_COUNT
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == deviceSeparatorPosition
                || position == boundedDeviceSeparatorPosition
                || position == availableDeviceSeparatorPosition) {
            ITEM_VIEW_TYPE_SEPARATOR
        } else {
            ITEM_VIEW_TYPE_REGULAR
        }
    }

    override fun isEnabled(position: Int): Boolean {
        return position != deviceSeparatorPosition
                && position != boundedDeviceSeparatorPosition
                && position != availableDeviceSeparatorPosition
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View? {
        val view: View
        val itemViewType = getItemViewType(position)
        view = if (convertView == null) {
            val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            if (itemViewType == ITEM_VIEW_TYPE_SEPARATOR) { // If separator
                inflater.inflate(R.layout.device_section_header, null)
            } else { // Regular row
                inflater.inflate(R.layout.device_item, null)
            }
        } else {
            convertView
        }
        if (itemViewType == ITEM_VIEW_TYPE_SEPARATOR) { // If separator
            val separatorView = view.findViewById(R.id.separator) as TextView
            if (position == deviceSeparatorPosition) {
                separatorView.text = deviceSeparatorText
            } else if (position == boundedDeviceSeparatorPosition
                    && boundedDeviceSeparatorPosition != 0) {
                separatorView.text = boundedDeviceSeparatorText
            } else if (position == availableDeviceSeparatorPosition
                    && availableDeviceSeparatorPosition != 0) {
                separatorView.text = availableDeviceSeparatorText
            }
        } else { // If regular row
            val deviceNameView = view.findViewById(R.id.device_name) as TextView
            val deviceAddressView = view.findViewById(R.id.device_address) as TextView
            // Set device name and address values
            if (position == 1) {
                deviceNameView.text = deviceName
                deviceAddressView.text = deviceAddress
            } else {
                val device: BluetoothDevice
                var positionInList = position - 3
                if (positionInList < boundedList.size) device = boundedList[positionInList] else {
                    positionInList -= boundedList.size + 1
                    device = availableList[positionInList]
                }
                deviceNameView.text = device.name
                deviceAddressView.text = device.address
            }
        }
        return view
    }

    fun addBoundedDevice(device: BluetoothDevice) {
        boundedList.add(device)
        updateSeparatorPositions()
        this.notifyDataSetChanged()
    }

    fun addAvailableDevice(device: BluetoothDevice) {
        availableList.add(device)
        this.notifyDataSetChanged()
    }

    private fun updateSeparatorPositions() {
        boundedDeviceSeparatorPosition = if (boundedList.size == 0) 0 else 2
        availableDeviceSeparatorPosition = if (availableList.size == 0) 0 else boundedList.size + boundedDeviceSeparatorPosition
    }

}